clients/kafka-common/src/test/java/org/eclipse/hono/client/kafka/consumer/AsyncHandlingAutoCommitKafkaConsumerTest.java

```diff
@@ -159,6 +160,76 @@ public void testConsumerCreationWithTopicPatternSucceeds(final VertxTestContext
+        final AtomicInteger offsetCounter = new AtomicInteger();
+        final int numTestBatches = 5;
+        final int maxPollRecords = 10;
+        final int maxRecordsInProcessing = maxPollRecords + 1;
+        final int numRecords = numTestBatches * maxPollRecords;
+        final Promise<Void> allRecordsReceivedPromise = Promise.promise();
+        final List<Promise<Void>> uncompletedRecordHandlingPromises = new ArrayList<>();
+        final List<KafkaConsumerRecord<String, Buffer>> receivedRecords = new ArrayList<>();
+        final AtomicInteger observedMaxRecordsInProcessing = new AtomicInteger();
+        final Function<KafkaConsumerRecord<String, Buffer>, Future<Void>> recordHandler = record -> {
+            receivedRecords.add(record);
+
+            final Promise<Void> recordHandlingCompleted = Promise.promise();
+            uncompletedRecordHandlingPromises.add(recordHandlingCompleted);
+            if (consumer.isPaused() || receivedRecords.size() == numRecords) {
+                if (uncompletedRecordHandlingPromises.size() > observedMaxRecordsInProcessing.get()) {
+                    observedMaxRecordsInProcessing.set(uncompletedRecordHandlingPromises.size());
+                }
+                uncompletedRecordHandlingPromises.forEach(Promise::tryComplete);
+                uncompletedRecordHandlingPromises.clear();
+            }
+            if (receivedRecords.size() == numRecords) {
+                vertx.runOnContext(v -> allRecordsReceivedPromise.tryComplete());
+            }
+            return recordHandlingCompleted.future();
+        };
+        final Map<String, String> consumerConfig = consumerConfigProperties.getConsumerConfig("test");
+        consumerConfig.put(ConsumerConfig.GROUP_ID_CONFIG, UUID.randomUUID().toString());
+        consumerConfig.put(ConsumerConfig.MAX_POLL_RECORDS_CONFIG, Integer.toString(maxPollRecords));
+
+        consumer = new AsyncHandlingAutoCommitKafkaConsumer(vertx, Set.of(TOPIC), recordHandler, consumerConfig);
+        consumer.setKafkaConsumerSupplier(() -> mockConsumer);
+        mockConsumer.updateEndOffsets(Map.of(topicPartition, ((long) 0)));
+        mockConsumer.updatePartitions(topicPartition, KafkaMockConsumer.DEFAULT_NODE);
+        mockConsumer.setRebalancePartitionAssignmentAfterSubscribe(List.of(topicPartition));
+        consumer.start().onComplete(ctx.succeeding(v2 -> {
+            // schedule the poll tasks
+            for (int i = 0; i < numTestBatches; i++) {
+                mockConsumer.schedulePollTask(() -> {
+                    for (int j = 0; j < maxPollRecords; j++) {
+                        final int offset = offsetCounter.getAndIncrement();
+                        mockConsumer.addRecord(new ConsumerRecord<>(TOPIC, PARTITION, offset, "key_" + offset, Buffer.buffer()));
+                    }
+                });
+            }
+            final long timerId = vertx.setTimer(8000, tid -> {
```

Comment:
is it really necessary to wait for this task for 8 seconds? Can't we use a mocked vert.x instance instead?

Reply:
The 8s here are just the timeout value (lower than the overall test timeout). Using this extra timer makes it possible to print some extra output in case this timeout occurs here.
And better to do this on the vert.x event loop thread. The alternative of using a `CountdownLatch` on the `main` thread seemed problematic - doing some things on the event loop thread and others on the `main` thread.

And concerning the vert.x instance, I think it wouldn't be easy to use a mocked instance with the `KafkaConsumer`. And it would introduce a change in behaviour in the KafkaConsumer compared with using a proper vert.x instance. So, I wouldn't want to use a mocked instance here.

Reply:
I see